Allow empty nonwrapping ranges

psarna opened the pull request:
Previous implementation of `nonwrapping_interval` (a.k.a. `nonwrapping_range`) did not handle empty intervals, which results in an unobvious interface - it's illegal(UB) to create an empty interval, e.g. `[0, 0)` or `[5, 1]`, and various functions return  `optional<nonwrapping_interval>` when the result might be empty - e.g. `intersection`. This miniseries adds support for empty intervals. It's RFC because the interface is still not pretty, but perhaps it's a step in the right direction.

Notes:
 * some of the tests operated on undefined ranges and even asserted certain results: e.g. `test_range_subtract`
 * `wrapping_interval` implements both `operator==` and `equal()`, where equal is the function that should be used for comparing intervals, since it accepts a comparator. `operator==` is quite useless, especially for now implemented empty ranges - `[5,1) == [6, 3)`, but it won't be so with current `operator==` implementation.
 * due to how intervals are implemented underneath, creating an empty `nonwrapping_interval<T>` needs a dummy value, since it's not always possible to create a default object of type `T`. In general, `empty` should be a bit field, but then we'd need to serialize it when sending it via network, which is not done now, otherwise the IDL layer won't be able to deduce if the interval is empty or not when it creates one. That's a bigger change though.

/cc @avikivity @dekimir

psarna commented on cql3/restrictions/statement_restrictions.cc:
```diff
@@ -1033,7 +1033,7 @@ struct intersection_visitor {
     value_set operator()(const nonwrapping_range<bytes>& a, const nonwrapping_range<bytes>& b) const {
         const auto common_range = a.intersection(b, type->as_tri_comparator());
-        return common_range ? *common_range : empty_value_set;
+        return !common_range.empty(type->as_tri_comparator()) ? std::move(common_range) : empty_value_set;
```
Hm, I didn't even consider it when coding this patch - it was just mechanical replacement from `std::optional` to calling `empty()`. It would be nice to simplify that in a follow-up though.

dekimir replied:
Yeah, a follow-up is the right place to do it.

avikivity replied:
Agree

psarna commented:
By the way, when performing further testing of this RFC, I found one more curiosity... I wanted to reorder some checks in the implementation of `deoverlap` - currently, the function starts with a check whether only one range was passed, and if so, it's already non-overlapping and we can return it. Later, empty intervals are erased, since they're not interesting wrt. overlapping, since they do not overlap with anything. But then, I figured that maybe I'll move the erasing step to the front of the function, and a test failed.
The test is `view_schema_ckey_test::test_clustering_key_multi_column_restrictions` and it's interesting because its materialized views have a filter. Now, it turned out that some ranges used in this case were incorrectly labelled as `empty`, even though they weren't... the problem is that if somebody calls `is_wrap_around()` on a singular range, it's possible that it will return `true`, which is obviously incorrect - the implementation of `is_wrap_around()` checks both bounds, but doesn't take into account that if `_singular` is checked, the `end` bound is unused and sholdn't be relied upon. I'll fix that and push v2, along with this improved `deoverlap` implementation.

psarna commented:
Hm, it's not that... singular ranges happen to work fine, because `_end` is disengaged, but still the range I get in the view code is a wrap-around... perhaps it's a bug in the materialized views path after all. They seem to rely on an UB nonwrapping range, in which end < start.

avikivity commented on interval.hh:
```diff
@@ -462,27 +459,35 @@ public:
+    bool empty(Comparator&& cmp) const {
+        return _interval.is_wrap_around(cmp);
+    }
+
```
Should we allow constructing ranges with lower > upper?

Options:
 1. yes
 2. disallow in constructor (throw or assert, or trust user), add separate free function that checks for the condition and constructs a real empty instead

psarna replied:
In order to disallow in constructor we need an access to a comparator in constructor - not always the case, especially with IDL.

avikivity replied:
I meant "logically disallow".

avikivity commented on interval.hh:
```diff
@@ -462,27 +459,35 @@ public:
     template<typename Comparator>
     bool before(const T& point, Comparator&& cmp) const {
-        return _interval.before(point, std::forward<Comparator>(cmp));
+        return !empty(cmp) && _interval.before(point, std::forward<Comparator>(cmp));
```
Having an internal _empty flag will speed up these operations.

psarna commented:
After some more debug prints to `view_schema_ckey_test::test_clustering_key_multi_column_restrictions`, I see that inside `calculate_affected_clustering_ranges`, a clustering range returned by `v->view_info()->partition_slice().default_row_ranges()` is invalid. The range is `ckp{00040000000000000001}, ckp{}`, which translates to a range starting with clustering key prefix with integer value `1` and ending with an empty prefix. Since an empty prefix compares less to any nonempty prefix, this range is actually degenerate - end < start, which is supposed to be disallowed by nonwrapping ranges. It's unfortunately not asserted, because constructors don't have access to comparators and can't check this.

Worse thing is that this degenerate range seems to mean something to the view update generation layer anyway - it's used for reading old base values. It suggests that the materialized views code relies on the undocumented fact that nonwrapping ranges are in fact wrapping underneath, and it sometimes works fine.

avikivity commented:
How did you check all the code? Where you have optional<interval> it's easy, but elsewhere?

I thought to rename all the interval methods that can give different result, let the compiler find build errors, audit each call site and rename it as it's fixed, then mass-rename back.

psarna commented:
My original assumption was that code which uses optional expects and handles empty ranges, while code without optionals already assumed that the ranges are never empty. Unfortunately, it appears that sometimes these degenerate "nonwrapping" ranges which are in fact wrapped, are used (see my comment above), and it would be rather hard to audit - changing names could give some nice starting points, but these ranges can get propagated to multiple call-sites later, so the only complete solution seems to be to carefully read every single line which uses range objects.

avikivity commented:
Do we have badly-wrapped ranges outside tests?

psarna commented:
Yes. `view_schema_ckey_test::test_clustering_key_multi_column_restrictions` is a test case which uses a higher layer (CQL), and it's possible to generate such a badly wrapped range in there.

psarna commented:
Unless I misunderstand what such a clustering range: `ckp{00040000000000000001}, ckp{}` is supposed to represent, but it seems to be a wrap-around range - it was definitely detected as such in my new code during deoverlapping, because `ckp{} < ckp{00040000000000000001}` in the order in which `clustering_key_prefix_view` objects are compared.

dekimir commented:
I just realized that this extends the functional overlap between nonwrapping_interval and value_set.  I wonder if we should instead switch to value_set in parts of the code that deal with non-token column ranges.

https://github.com/scylladb/scylla/blob/29fccd76eaaa0da641b91c8c64bd4957edbe5c9f/cql3/restrictions/restriction.hh#L127-L132

avikivity commented:
> Unless I misunderstand what such a clustering range: `ckp{00040000000000000001}, ckp{}` is supposed to represent, but it seems to be a wrap-around range - it was definitely detected as such in my new code during deoverlapping, because `ckp{} < ckp{00040000000000000001}` in the order in which `clustering_key_prefix_view` objects are compared.

This is supposed to be empty (and if the user provides such bounds we need to fold those inputs to an empty range).

Note that this is different from `ckp{blah}, nullopt` which is +Inf on the right.

psarna commented:
I'm pretty sure that `ckp{00040000000000000001}, ckp{}` was not provided by the user, but by the materialized view code. It **should** represent an empty range, but it definitely does not in our current code, because the new code, which purged empty ranges during `deoverlap`, broke a test case. Now that I'm back from vacation, I'll dig further and provide some concrete steps on what's wrong and how to reproduce it, instead of just hand-waving.

psarna commented:
The degenerate range is created due to `(b, c)  >= (1, 0)` restriction residing in the materialized view. Namely, the code below:
```cpp
    std::vector<bounds_range_type> bounds_ranges_unified_order(const query_options& options) const {
        auto start_prefix = clustering_key_prefix::from_optional_exploded(*_schema, read_bound_components(options, statements::bound::START));
        auto start_bound = bounds_range_type::bound(std::move(start_prefix), _slice.is_inclusive(statements::bound::START));
        auto end_prefix = clustering_key_prefix::from_optional_exploded(*_schema, read_bound_components(options, statements::bound::END));
        auto end_bound = bounds_range_type::bound(std::move(end_prefix), _slice.is_inclusive(statements::bound::END));
        auto make_range = [&] () {
            if (is_asc_order()) {
                return bounds_range_type::make(start_bound, end_bound);
            } else {
                return bounds_range_type::make(end_bound, start_bound);
            }
        };
        auto range = make_range();
        auto bounds = bound_view::from_range(range);
        if (bound_view::compare(*_schema)(bounds.second, bounds.first)) {
            return { };
        }
        return { std::move(range) };
    }
```
does not seem to have a notion of open-ended ranges - it always creates a bound, while an open-ended range is represented by having its `optional<bound>` set to nullopt. As a result, `(b, c)  >= (1, 0)` is parsed to `ck >= ckp{00040000000000000001} and ck <= ckp{}` instead of `ck >= ckp{00040000000000000001} and ckp{} <= +Inf`.
It looks like the code works fine just because it piggybacks on the undocumented fact that nonwrapping ranges are kind-of wrapping, and once I removed that behaviour, the test broke. I'm a little bit overloaded today since I'm cathing up from last week, but I'll fix this quirk in v2 of this series.

psarna commented:
v2 changes:
 - rebased
 - added a fix for incorrect ranges generated for multi-column restrictions
 - changed deoverlap to always try to normalize-out empty ranges first

avikivity commented on interval.hh:
```diff
@@ -426,7 +426,7 @@ std::ostream& operator<<(std::ostream& out, const wrapping_interval<U>& r) {
 
 // An interval which can have inclusive, exclusive or open-ended bounds on each end.
-// The end bound can never be smaller than the start bound.
+// The end bound smaller than the start bound indicates an empty interval.
```
I think it's not good to rely on this. Empty intervals can be created with the default constructor, and it should be illegal to call an interval with swapped parameters.

avikivity commented:
I queued the first two patches. I'm still worried about the API for empty intervals, and whether they can now appear in surprising places.

psarna commented:
Yup, it is worrisome. I wonder if a similar trick to what we did for tokens could be applied... in order to be able to compare them without context and allow per-table partitioners, we simply deprecated non-64bit partitioners and used the fact that all tokens are now 64bit. Here, it's not possible to compare two bounds without context (a comparator), but perhaps it's not needed? E.g. even if we don't know how to compare `X` and `Y`, we do know that a range `[X, X)` is for sure empty, regardless of what `X` is. I'm just waving hands again, I'll think more about it tomorrow. We still won't be able to verify if a range is correct when constructing it from a web packet (unless we provide more context in the packet itself), but, still, we can check correctness later, and use `[X, X)` as the only valid representation of an empty interval - then, no additional fields (like an `is_empty` bit) would need to be added to our protocols, which would help maintain backward compatibility. (On the other hand, an `is_empty` bit would make the checks faster, without having to endlessly compare identical values, so it can be cached inside the interval after the first full check).

avikivity commented:
I think an is_empty bit is useful as it makes a default constructor sensible and possible.

But my main worry is that we well have users of ranges that previously did not expect empty ranges^Wintervals, and would now get them. But maybe that's impossible - the compiler made you audit all all sites of intersection(), and you fixed up deoverlap(). So the only source of empty intervals is constructors, and those weren't protected even before. So I think the patch is actually safe.

slivne commented:
@avikivity / @psarna is this still relevant - if not lets close

psarna commented:
Well, the first two patches fix bugs, so it would be nice to have them in, even if the third one is considered too controversial. I'll leave the decision to the reviewers though.

psarna commented:
Rebased for potential re-review

avikivity reviewed (commented):
> Rebased for potential re-review

Same evaluation. It's logically the correct thing to do, but need a full audit of the code to make sure we either don't pass an empty interval to code that doesn't expect it, or that we check for empty intervals everywhere.

I suppose the latter is safer.

One way to go about it:
 - duplicate every accessor foo into a reviewed_foo(). foo() is marked deprecated and calls reviewed_foo().
 - build the code base, look at every call site, add checks for empty, change accessors to reviewed_foo()
 - repeat until the build is clean or you die of boredom
 - extract the new emptiness checks, drop reviewed_foo
